core/arch/arm/kernel/generic_entry_a32.S

```diff
@@ -213,11 +213,16 @@ UNWIND(	.cantunwind)
 	ldr	r0, =__text_init_start	/* dst */
 	ldr	r1, =__data_end 	/* src */
 	ldr	r2, =__tmp_hashes_end	/* dst limit */
+	/* Copy backwards (as memmove) in case we're overlapping */
+	sub	r2, r2, r0		/* len */
+	add	r0, r0, r2
+	add	r1, r1, r2
+	ldr	r2, =__text_init_start
 copy_init:
-	ldm	r1!, {r6-r12}
-	stm	r0!, {r6-r12}
+	ldmdb	r1!, {r6-r12}
```

Comment:
Yes, we may very well copy a few more than needed. But that should be harmless because we are supposed to have a quite large space of unused memory surrounding. We'll probably write a few bytes inside the area that soon will be initialized for heap or translation tables.